Combined Metrics API updates for v0.3 (Add Observer instrument, remove Gauge instrument, etc.) (#430)
open-telemetry/opentelemetry-specification: specification/api-metrics.md

# Metrics API

<!-- toc -->

- [Overview](#overview)
  * [User-facing API](#user-facing-api)
  * [Meter API](#meter-api)
  * [Purpose of this document](#purpose-of-this-document)
- [Metric API / SDK separation](#metric-api--sdk-separation)
  * [Three kinds of instrument](#three-kinds-of-instrument)
    + [Counter](#counter)
    + [Measure](#measure)
    + [Observer](#observer)
  * [Standard Interpretation](#standard-interpretation)
  * [Optional semantic restrictions](#optional-semantic-restrictions)
    + [Absolute vs. Non-Absolute](#absolute-vs-non-absolute)
    + [Monotonic vs. Non-Monotonic](#monotonic-vs-non-monotonic)
- [Metric instrument selection](#metric-instrument-selection)
  * [Counters and Measures compared](#counters-and-measures-compared)
  * [Observer instruments](#observer-instruments)
- [Examples](#examples)

<!-- tocstop -->

## Overview

> The OpenTelemetry Metrics API supports producing diagnostic

Reviewer: I think introductions could be extended a bit. Consider the following notions:
  * all telemetry APIs are ultimately designed to capture events. What makes the Metrics API different is that it is explicitly designed for _measuring_ things (capturing observations) and producing _aggregations_, rather than raw events.
  * the output of the Metrics API are time series of observations, with or without(?) aggregation
  * a "metric" (I am not sure about this being the best term, what about "time series"?) should have a _definition_, which may include a description, unit of measure (temperature in F vs. C, request size in bytes, etc.), data validation policy (monotonic, non-negative, how to treat non-conformant values), and maybe even desired aggregation
  * an instrument is always created from a definition

Author: > all telemetry APIs are ultimately designed to **capture**...

I had debated (with myself) using "capture" (_take into one's possession or control by force_) as the primary verb, instead I used "produce" (_make or manufacture from components or raw materials_). 

> the output of the Metrics API are time series of observations, with or without(?) aggregation

I intentionally avoided "time series" in this document (as you noticed), because it seemed to have an association with the output format. This is a word that exporters use in their exposition format, not necessarily a semantic term, I think. Thus far, time has been an implicit property in the API. Calling these "time series" makes that association more explicit, I think. Is the API required to compute the current timestamp on every API call? (I would say not). Is the required to order the series of events by timestamp? (I would say not). Do we allow users to provide their own timestamps? (I would say not). All of these questions lead me away from the term "time series".

I realize now that the specification has to clear this up and mention time more explicitly.

I wasn't careful about introducing the term "event", but I think it's more general than "observations". I would say that Observer instruments record a set of observations as a single event ("capture" might be used here, i.e., Observer instruments capture a set of observations).

On the other hand, I would say that Counter and Measure instruments are used to record observations made by the user of the API ("capture" sometimes applies here, i.e., these instruments are used to record observations captured by the user).

> a "metric" (I am not sure about this being the best term...

FWIW, I've been careful to avoid using "metric" as a noun. It's too vague, so I've used it as an adjective as in "metric instrument". 

> should have a definition ...

In an earlier draft, I had defined a "descriptor" which encapsulated the details provided during instrument definition. @bogdandrutu asked if it was necessary [here](https://github.com/open-telemetry/opentelemetry-specification/pull/299#discussion_r333789441). 

> an instrument is always created from a definition

I have been writing "an instrument is defined" to avoid defining a "metric definition" or descriptor type, and have been referring to properties of the instrument instead (e.g., "the instrument name"). I am inclined to re-introduce this concept.

Author: I've rewritten the introduction to this document, everything through line 200 or so. I switched to using "capture", introduce "metric event" explicitly very early in the document, add the term "instrument definition", and discuss time more carefully.

used by the program to record observations about their behavior.
Therefore, we use "metric instrument" to refer to a programmatic
interface, allocated through the API, used to produce metric events.
There are three kinds of instruments known as Counters, Measures, and
Observers.

Monitoring and alerting are the common use-case for the data provided
through metric instruments, after various collection and aggregation
strategies are applied to the data.  We find there are many other uses
for the _metric events_ recorded through these instruments.  We
imagine metric data being aggregated and recorded as events in tracing
and logging systems too, and for this reason OpenTelemetry requires a
separation of the API from the SDK.

### User-facing API

The user-facing OpenTelemetry API for Metrics begins with a `Meter`
interface, usually obtained through dependency injection or a global
instance.  The `Meter` API supports defining new metric instruments.
Review the [user-facing OpenTelemetry API
specification](api-metrics-user.md) for more detail about the variety
of methods, options, and optimizations available for users of the

> instrumentation API and how to use the instruments defined here.

Author: I want this to go in the companion document, still un-updated. (api-metric-user.md)

### Meter API

To produce measurements using an instrument, you need an SDK that

> supports the `Meter` API, which consists of a set of constructors, the

Author: Done.

`Labels` function for building label sets, and the `RecordBatch`
function for batch reporting.  Refer to the [sdk-facing OpenTelemetry
API specification](api-metrics-meter.md) for more implementation notes.

Because of API-SDK separation, the `Meter` implementation ultimately
determines how metrics events are handled.  The specification's task
is to define the semantics of the events in high-level terms, so that
users and implementors can agree on their meaning.  How the `Meter`
accomplishes its goals and the export capabilities it supports are
specified for the default SDK in the (Metric SDK specification

> WIP)[#WIP-spec-issue-347].

Author: Fixed.

> The standard implementation for `Meter` implementations to follow is

Author: Yes. This paragraph is gone, but I've tried to improve the explanation of this idea.

given, to aid with understanding the different instrument use-cases.
For example, a Counter instrument supports `Add()` events, and the
standard implementation is to compute a sum.  The sum may be exported
as a grand total or as the change in the grand total, but regardless
of the exporter and specific techniques, the purpose of using a
Counter with `Add()` is to monitor a sum.  A detailed explanation for
how to select metric instruments for common use-cases is given below,
according to the semantics defined next.

### Purpose of this document

This document gives an overview of the specification, introduces the
the three kinds of instrument, and discusses how end-users should
think about various instruments and options at a high-level, without
getting into detail about specific function calls.  For details about
specific function calls, refer to the detailed specifications linked

> above.

Author: Removed.

## Metric API / SDK separation

The API distinguishes metric instruments by semantic meaning, not by
the type of value produced in an exporter.  This is a departure from
convention, compared with a number of common metric libraries, and
stems from the separation of the API and the SDK.  The SDK ultimately
determines how to handle metric events and could potentially implement
non-standard behavior.  All metric events can be represented as

> consisting of a timestamp, an instrument, a number (the _value_), and

Author: Done.

a label set.  The semantics defined here are meant to assist both
application and SDK implementors, and examples will be given below.

The separation of API and SDK explains why the metric API does not
have metric instruments that generate specific metric "exposition"
values, for example "Histogram" and "Summary" values, which are
different ways to expose a distribution of measurements.  This API
specifies the use of a Measure kind of instrument with `Record()` for

> recording individual measurements.  The instruments are defined so

Reviewer 2: It seems like this sentence is intended to contrast the previous one describing how the API intends to acomplish recording a distribution of measurements. I think stating this explicitly helps continuity.

```suggestion
recording individual measurements of a distribution instead.  The instruments are defined so
```

> that users and implementors understand what they mean, beacuse

Reviewer 3: ```suggestion
that users and implementors understand what they mean, because
```

> that users and implementors understand what they mean, beacuse

Author: Done.

different SDKs will handle events differently.

> There is a common metric instrument known as a "Gauge" that is not

Reviewer 2: This is a run-on sentence. Can we refactor to something like the following?

> There is a common metric instrument known as a "Gauge" that is not included in this API. A Gauge instrument is one that reports the last value of a measuring device (e.g., a speedometer on your car's dashboard).

included in this API, the term "Gauge" referring to an instrument,
often mechanical, for reading the current "last" value of a measuring
device (e.g., a speedometer on your car's dashboard).  The problem
with "Gauge" starts from the term itself, which is figurative in

> nature.  Using the word "gauge" suggests a behavior, that the

Reviewer 2: Do we want to include "suggesting a behavior" as reason to exclude? It seems the same reason would exclude Counter.

instrument will be used to expose the last value, not a semantic
definition.  Uses of traditional gauge instruments translate into uses

> of Observer or Measure instruments in this API.

Reviewer 2: I agree this probably belongs somewhere other than this section.

> of Observer or Measure instruments in this API.

Author: Done.

### Label sets

_Label_ is the term used to refer to a key:value attribute associated
with a metric event.  Although they are fundamentally similar to [Span
attributes](api-tracing.md#span) in the tracing API, a set of labels
is given its own type in the Metric API, `LabelSet`.  Label sets are a
feature of the API to facilitate re-use across the metric API.  Users
are encouraged to re-use label sets whenever possible, as they may
contain a previously encoded representation of the data.

Users obtain label sets by calling the `Meter` API function (e.g.,
`Meter.GetLabels({ key, value }, ...)`).  Each of the instrument
calling conventions detailed in the [user-facing API
specification](api-metrics-user.md) accepts a `LabelSet`.

### Three kinds of instrument

#### Counter

Counter instruments are used to report sums.  These are commonly used
to monitor rates, and they are sometimes used to report totals.  One
key property of Counter instruments is that two `Add(1)` events are
semantically equivalent to one `Add(2)` event--`Add(m)` and `Add(n)`
is equivalent to `Add(m+n)`.  This means that Counter events can be

> combined using addition by definition.

Reviewer 2: We might want to drop this last sentence. It might confuse readers to interpret that an implementation of a counter in a language supports the addition operation of the language.

Author: Fixed.

Labels associated with Counter instrument events can be used to
compute rates and totals over selected dimensions.  When aggregating
Counter events we naturally combine values using addition.  Counter
`Add(0)` events are no-ops by definition.

Counters are monotonic by default, meaning `Add()` logically accepts

> only non-negative values.  Monotonicity is useful for defining rates,

Reviewer 2: nit: Counters are used to measure accumulations of events that can be interpreted as rates by looking at their derivative. It is not (to my knowledge) common to directly report rates with Counters.  I think it is fair in stating that they can be used to "monitor rates" (as is done in the beginning of this section), but this looseness seems to bleed into language like this that talks about defining rates directly.

especially; non-monotonic Counter instruments are an option to support
sums that rise and fall.

Examples: requests processed, bytes read or written, memory allocated

> and deallocated.

Author: Fixed. / Added.

#### Measure

Measure instruments are used to report individual measurements.
Measure events are semantically independent and cannot be naturally
combined like with Counters.  Measure instruments are used to report

> many kinds of information, recommended for reporting measurements

Reviewer 2: Is this missing an `and are`?

```suggestion
many kinds of information, and are recommended for reporting measurements
```

Author: Done.

associated with real events in the program.

As a synchronous API, Measure `Record()` events can be used to record
information with associated labels and context.  This is the more

> general of the two synchronous metric instruments.

Reviewer 2: Should we call out in the Counter section that it is classified as a `synchronous` metric as well? It seems implied the reader understands what the other synchronous metric is at this point.

Author: Done. This is now mentioned above, too.

> Measure instruments support non-negative values by default, also known

Reviewer 2: ```suggestion
Measure instruments support only non-negative values by default, also known
```

> Measure instruments support non-negative values by default, also known

Author: I removed option support as discussed in todays SIG meeting.

as "absolute" in the sense that, mathematically, absolute values are
never negative.  As an option, Measure instruments can be defined as
not absolute, supporting both postive and negative values.

Examples: request latency, number of query terms, temperature, fan
speed, account balance, load average, screen width.

#### Observer

Observer instruments are used to report a current set of values at the
time of collection.  Observer instruments report not only current
values, but also _which label sets are current_ at the moment of
collection, as a coherent set of values.  These instruments reduce
collection cost because they are computed and reported only once per

> collection interval, by definition.

Reviewer 4: Also only if ever required, for example if an implementation allows to enable/disable observer metrics we avoid querying the source of the metric - another optimization compared with Gauge.

Author: I added 
```
Because Observer instruments are activated by the SDK, they can be effectively disabled at low cost.
```

Unlike Counter and Measure instruments, Observer instruments are

> synchronized with collection, used to report values, on demand,

Reviewer 2: This seems like a run-on sentence.

```suggestion
synchronized with collection. They are used to report values, on demand,
```

calculated by the program itself.  There is no aggregation across time

> for Observer instruments by definition, only the current value is

Reviewer 2: Do we want to use "last value" instead of "current value" here? It might help convey the replacement value of the Observer for the Gauge.

defined.

> Observer instruments can be declared as monotonic.  A monotonic

Author: I removed option support as discussed in todays SIG meeting.

form a sufficient basis for expressing nearly all metric data.  But if
the API and SDK are separated, and the SDK can handle any metric event
as it pleases, why not have just one kind of instrument?  How are the
instruments fundamentally different, despite all metric events having
the same form (i.e., a timestamp, an instrument, a number, and a label
set)?

Establishing different kinds of instrument is important because in
most cases it allows the SDK to provide good default functionality
without requiring alternative behaviors to be configured.  The choice
of instrument determines not only the meaning of the events but also
the name of the function used to report data.  The function
names--`Add()` for Counter instruments, `Record()` for Measure
instruments, and `Observe()` for Observer instruments--help convey and
reinforce the standard interpretation of the event.

The standard implementation for the three instruments is defined as
follows:

1. Counter.  The `Add()` function accumulates a total for each
distinct label set.  When aggregating over distinct label sets for a
Counter, combine using addition.  Export as a set of calculated sums.

> 2. Measure.  Use the `Record()` function to report summary statistics

Reviewer 2: The `Record()` function is not used to "report summary statistics about the distribution of values" it is used to "record the individual measured value of an event", right? 

The summary statistics are the analytics generated by some aggregation.

Author: Fixed.

about the distribution of values, for each distinct label set.  Which

> statistics are used is determined by the implementation, but they

Author: I changed to "determined by the aggregation". I documented that configurable aggregations is future work.

usually include at least the sum of values, the count of measurements,
and the minimum and maximum values.  When aggregating distinct label
sets for a Measure, report summary statistics of the combined value
distribution.  Exposition formats for Measure data vary widely by
backend service.
3. Observer.  Current values are provided by the Observer callback at
the end of each Metric collection period.  When aggregating values
_for the same label set_, combine using the most-recent value.  When

> aggregating values _for different label sets_, combine using the sum.

Author: I've written that Observer instruments behave like Measure instruments, with a caveat. The guidance for users that want an additive observer will be to read the distribution sum, for users that want an average observer, read the distribution mean.

Export as pairs of label set and calculated value.

We recognize that the standard behavior of the three instruments does
not cover all use-cases perfectly.  There is a natural tension between
offering dedicated metric instruments for every distinct metric
application and combining use-cases, generalizing semantics to reduce
the API surface area.  We could have define more than or fewer than
three kinds of instrument; we have three because these seem like
enough.  Where uncommon use-cases call for a non-standard
implementation configuration (e.g., a Measure instrument configured
with last-value aggregation), we accept that users will be required to

> provide additional input on how to view certain metric data.

Reviewer 5: As @tsloughter [points out](https://github.com/open-telemetry/oteps/pull/80#issuecomment-577333331), making Counter and Observer top-level API constructs sounds like a great idea, in the spirit of making commonplace things easy.

But it is still pretty surprising to see Gauge and Counter classes in the guts of the SDK implementation instead of treating these as optimizations, or aliases for certain Measure configurations. For example, the `AggregatorSelector` in go serves to set a default aggregation for each metric instrument type, but each `AggregatorSelector` [returns the same aggregation for Counters](https://github.com/open-telemetry/opentelemetry-go/blob/6051c8149381d24d8a872035b34b988b28b7832c/sdk/metric/selector/simple/simple.go#L76) because Counters are always sum-aggregated. Would we have written it this way if we'd started with Measurements and added other (API) instrument types later?

Reviewer 4: @c24t is your question related to how the implementation is done, or how the API looks like? I think this document is about the API only.

Reviewer 5: I think this is a case of the API spec bleeding into the implementation. If (e.g.) the doc were written like this, I think we'd see different implementation choices in the prototypes:

The most general form of metric instrument is a Measure. By default, a Measure records summary statistics about the distribution of values for each distinct labelset. A user may configure a Measure to be aggregated in accordance with a particular exposition format. Exposition formats for measure data may vary widely by backend, so the set of available aggregators is determined by the SDK.

The metrics API also provides a dedicated metric instrument for reporting values that should be summed. This instrument is a called a Counter, and is typically used to compute rates and counts. Measurements from counters are always aggregated by sum.

Reviewer 5: One hidden assumption here might be that all measures map to the same backend metric type. Do we want to make it possible to (e.g.) report some measures as histograms to prometheus and others as summaries?

Author: I put in a "future work" section about configurable aggregation.

Author: > Gauge and Counter classes in the guts of the SDK implementation

I disagree -- all the synchronous instruments should look the same by the time an SDK sees an event. Inside the implementation we have aggregators that treat the different events differently.

inputs, whereas non-absolute instruments can be described as accepting
signed inputs.

When an instrument is absolute (i.e., accepts non-negative updates),
we know that the sum can be used to express a rate automatically.
This is true for all kinds of instrument.

When exporting measure values as a histogram, for example, knowing the
instrument is absolute facilitates the use of logarithmic buckets
(which are difficult to use when the input range spans zero).

Absolute behavior is the default for all instrument kinds. The
Non-Absolute option is supported for all instrument kinds.

Because this is a simple property for the SDK to test, the
specification recommends that SDKs SHOULD reject metric events for
absolute instruments when negative values are used, and instead issue
a warning to the user.

#### Monotonic vs. Non-Monotonic

Monotonic refers to whether an instrument only accepts values that are
greater than or equal to the previously recorded value.  Non-monotonic
instruments are those which accept any change in the value, positive
or negative.

Absolute-valued counters are naturally monotonic, so that Absolute and

> Monotonic have the same interpretation for Counter instruments.

Author: I removed all option support as discussed in the SIG meeting.

Measure and Observer instruments may be declared as monotonic, however
since this property is expensive to test, the specification recommends
that SDKs SHOULD implement monotonicity checking only when computing a
last-value aggregation.  The SDK SHOULD only perform this test against
the last known value when it holds the necessary information, it
should not go out of its way to save data simply to perform
monotonicity testing.

## Metric instrument selection

To guide the user in selecting the right kind of metric instrument for
an application, we'll consider several questions about the kind of
numbers being reported.  Here are some ways to help choose.  Examples
are provided in the following section.

### Counters and Measures compared

Counters and Measures are both recommended for reporting measurements
taken during synchronous activity, driven by events in the program.
These measurements include an associated distributed context, the
effective span context (if any), the correlation context, and

> user-provided LabelSet values.

Reviewer: maybe the sensitivity to the context should be mentioned earlier.

Author: Will do.

report any kind of measurement that is not typically expressed as a
rate or as a total sum.

If the Measure instrument accepts only non-negative values, as is
typically the case for measuring physical quantities, use a (default)
absolute Measure.  If the Measure instrument accepts both positive and
negative values, use a non-absolute Measure.  Both of these are
typically expressed in terms of a distribution of values, independent
from and _in addition to_ the rate of these measurements.

### Observer instruments

Observer instruments are recommended for reporting measurements about
the state of the program at a moment in time.  These expose current
information about the program itself, not related to individual events
taking place in the program.  Observer instruments are reported
outside of a context, thus do not have an effective span context or
correlation context.

Observer instruments are meant to be used when measured values report
on the current state of the program, as opposed to a change of state
in the program.  When Observer instruments are used to report physical
quantities, use a (default) absolute Observer.  When Observer

> instruments are used to report measurements can be negative for any

Reviewer: ```suggestion
instruments are used to report measurements that can be negative for any
```

reason, use a non-absolute Observer.

If the Observer reports a current total sum, declare it as a monotonic
Observer.  Monotonic values are typically expressed as a rate of
change.

## Examples

### Reporting bytes read and written

You wish to monitor the number of bytes read and written from a
messaging server that supports several protocols.  The number of bytes
read and written should be labeled with the protocol name and
aggregated in the process.

This is a typical application for the Counter instrument.  Use one
Counter for bytes read and one Counter for bytes written.  When
handling a request, compute a LabelSet containing the name of the
protocol and potentially other useful labels, then call `Add()` twice
with the same label set and the number of bytes read and written.

> To make lower the cost of this reporting, you can `Bind()` the

Reviewer: ```suggestion
To lower the cost of this reporting, you can `Bind()` the
```

instrument with each of the supported protocols ahead of time and
avoid computing the label set for each request.

### Reporting per-request CPU usage

Suppose you have a way to measure the CPU usage of processing an
individual request.  This is given to you in terms of cpu-seconds
consumed.  You may wish to monitor total CPU usage, or you could be
interested in the peak rate of CPU usage.

> Use a Counter instrument to `Add()` this quantity to an instrument

Reviewer: What if we used the term Sum instead of Counter? It's even more explicit semantically, and goes equally well with integer and float values, whereas Counter that takes float values is kind of weird.

Author: I like this suggestion. "Counter" is itself a noun and sounds like an instrument (i.e., _an apparatus used for counting_), so there's redundancy there. "Counter" is also conventional in both Prometheus and Statsd.

On the other hand, while "Sum" isn't commonly used as an adjective, it describes the semantics of the instrument perfectly.

The "Measure" in Measure instrument is also a semantic adjective, and now I'm not sure that "Observer" instrument fits this trend. I considered whether these could be called "Instantaneous instruments", because they report a set of values at some instant. That's a lot of letters. The thing Observers do is create instantaneous measurements. I'm interested in what others think about the terms "Sum" and "Instantaneous" (or other suggestions).

named `cpu.seconds.used` after sending the response.  A Counter is
called for, in this case, because a sum is requested, meaning a sum of
all `Add()` events for the instrument in the specified time range.

### Reporting system call duration

You wish to monitor the duration of a specific system call being made
frequently in your application, with a label to indicate a file name
associated with the operation.

This is a typical application for the Measure instrument.  Use a timer
to measure the duration of each call and `Record()` the measurement
with a label for the file name.

### Reporting request size

You wish to monitor a trend in request sizes, which means you are
interested in characterizing individual events, as opposed to a sum.
Label these with relevant information that may help explain variance
in request sizes, such as the type of the request.

This is a typical application for a Measure instrument.  The standard
aggregation for Measure instruments will compute a measurement sum and
the event count, which determines the mean request size, as well as

> the minimum and maximum sizes.

Reviewer: may want to add an example of partitioning by correlation context dimensions.

Author: Will do.

Author: ```
### Reporting bytes read and written correlated by end user

An application uses storage servers to read and write from some
underlying media.  These requests are made in the context of the end
user that made the request into the frontend system, with Correlation
Context passed from the frontend to the storage servers carrying these
properties.

Use Counter instruments to report the number of bytes read and written
by the storage server.  Configure the SDK to use a Correltion Context
label key (e.g., named "app.user") to aggregate events by all metric
instruments.
```

# Metrics API

<!-- toc -->

- [Overview](#overview)
  * [User-facing API](#user-facing-api)
  * [Meter API](#meter-api)
  * [Purpose of this document](#purpose-of-this-document)
- [Metric API / SDK separation](#metric-api--sdk-separation)
  * [Three kinds of instrument](#three-kinds-of-instrument)
    + [Counter](#counter)
    + [Measure](#measure)
    + [Observer](#observer)
  * [Standard Interpretation](#standard-interpretation)
  * [Optional semantic restrictions](#optional-semantic-restrictions)
    + [Absolute vs. Non-Absolute](#absolute-vs-non-absolute)
    + [Monotonic vs. Non-Monotonic](#monotonic-vs-non-monotonic)
- [Metric instrument selection](#metric-instrument-selection)
  * [Counters and Measures compared](#counters-and-measures-compared)
  * [Observer instruments](#observer-instruments)
- [Examples](#examples)

<!-- tocstop -->

## Overview

The OpenTelemetry Metrics API supports capturing measurements about
the execution of a computer program in real time.  The Metrics API is
designed explicitly for processing raw measurements, generally with

> the intent to produce continuous summaries those measurements, also in

Author: Done.

sums that rise and fall.

Examples: requests processed, bytes read or written, memory allocated
and deallocated.

### Measure

Measure instruments are used to report individual measurements.
Measure events are semantically independent and cannot be naturally
combined like with Counters.  Measure instruments are used to report
many kinds of information, recommended for reporting measurements
associated with real events in the program.

As a synchronous API, Measure `Record()` events can be used to record
information with associated labels and context.  This is the more
general of the two synchronous metric instruments.

Measure instruments support non-negative values by default, also known
as "absolute" in the sense that, mathematically, absolute values are
never negative.  As an option, Measure instruments can be defined as
not absolute, supporting both postive and negative values.

Examples: request latency, number of query terms, temperature, fan
speed, account balance, load average, screen width.

### Observer

Observer instruments are used to report a current set of values at the
time of collection.  Observer instruments report not only current

> values, but also _which label sets are current_ at the moment of

Author: Not sure I follow?

> values, but also _which label sets are current_ at the moment of

Reviewer 6: I mean the label values for observer would be known at the time of its creation and would not change when it is collected. This line `Observer instruments report not only current
values, but also _which label sets are current_ at the moment of collection` indicates that there could be different labels between two collections.

> values, but also _which label sets are current_ at the moment of

Author: Yes, that is what I expect. Do you think these instruments should be required to use only a static set of label sets? I think that might be overly restrictive.

> values, but also _which label sets are current_ at the moment of

Reviewer 6: I would prefer to be restrictive for now until we have a proper use-case otherwise.

> values, but also _which label sets are current_ at the moment of

Author: There are already examples in this document that imply a dynamic set of labels can be produced, such as the one about exporting per-shard memory holdings. If the set of shards is dynamic, so are the observations.

All of the other examples I can think of are to export a distribution of current sizes. E.g., you have an index containing spans and have a count by distinct span names. Your observations are dynamic depending on the current set of distinct span names.

> values, but also _which label sets are current_ at the moment of

Reviewer 6: I mis-interpreted current_ labels as labels coming from current context. The current context would be of the collector observing the values. But that is not the case. Instead it reports  multiple current values each associated with unique combination of label values.

> values, but also _which label sets are current_ at the moment of

Author: Ah! I realize I had omitted an important detail, that Observer events have no associated Context.

> values, but also _which label sets are current_ at the moment of

Author: I added 
```
This means, for example, it is not possible
to associate Observer instrument events with Correlation or Span
context.
```

instruments, and `Observe()` for Observer instruments--help convey and
reinforce the standard interpretation of the event.

The standard implementation for the three instruments is defined as
follows:

1. Counter.  The `Add()` function accumulates a total for each
distinct label set.  When aggregating over distinct label sets for a
Counter, combine using addition.  Export as a set of calculated sums.
2. Measure.  Use the `Record()` function to report summary statistics
about the distribution of values, for each distinct label set.  Which
statistics are used is determined by the implementation, but they
usually include at least the sum of values, the count of measurements,
and the minimum and maximum values.  When aggregating distinct label
sets for a Measure, report summary statistics of the combined value
distribution.  Exposition formats for Measure data vary widely by
backend service.
3. Observer.  Current values are provided by the Observer callback at
the end of each Metric collection period.  When aggregating values
_for the same label set_, combine using the most-recent value.  When
aggregating values _for different label sets_, combine using the sum.

> Export as pairs of label set and calculated value.

Author: Done.

Measure and Observer instruments may be declared as monotonic, however
since this property is expensive to test, the specification recommends
that SDKs SHOULD implement monotonicity checking only when computing a
last-value aggregation.  The SDK SHOULD only perform this test against
the last known value when it holds the necessary information, it
should not go out of its way to save data simply to perform
monotonicity testing.

## Metric instrument selection

To guide the user in selecting the right kind of metric instrument for
an application, we'll consider several questions about the kind of
numbers being reported.  Here are some ways to help choose.  Examples
are provided in the following section.

### Counters and Measures compared

Counters and Measures are both recommended for reporting measurements
taken during synchronous activity, driven by events in the program.
These measurements include an associated distributed context, the
effective span context (if any), the correlation context, and
user-provided LabelSet values.

Start with an application for metrics data in mind.  It is useful to
consider whether you are more likely to be interested in the sum of

> values or the average value, as processed by the instrument.  Counters

Author: Done.

values are of interest.

If only the sum is of interest, use a Counter instrument.  If the
Counter instrument accepts non-negative `Add()` values, use a
(default) monotonic Counter which will typically be expressed as a
rate (i.e., change per unit time).  If the Counter accepts both
positive and negative `Add()` values, use a non-monotonic Counter
which will typically be expressed as the total sum.

If you are interested in any other kind of summary value or statistic,
such as mean, median and other quantiles, or minimum and maximum
value, use a Measure instrument.  Measure instruments are used to
report any kind of measurement that is not typically expressed as a
rate or as a total sum.

If the Measure instrument accepts only non-negative values, as is
typically the case for measuring physical quantities, use a (default)
absolute Measure.  If the Measure instrument accepts both positive and
negative values, use a non-absolute Measure.  Both of these are
typically expressed in terms of a distribution of values, independent
from and _in addition to_ the rate of these measurements.

### Observer instruments

Observer instruments are recommended for reporting measurements about

> the state of the program at a moment in time.  These expose current

Author: Done.

  * [Standard implementation](#standard-implementation)
  * [Option: Dedicated Measure instrument for timing measurements](#option-dedicated-measure-instrument-for-timing-measurements)
  * [Future Work: Option Support](#future-work-option-support)
  * [Future Work: Configurable Aggregations / View API](#future-work-configurable-aggregations--view-api)
- [Metric instrument selection](#metric-instrument-selection)
  * [Counters and Measures compared](#counters-and-measures-compared)
  * [Observer instruments](#observer-instruments)
- [Examples](#examples)
  * [Reporting bytes read and written](#reporting-bytes-read-and-written)
  * [Reporting per-request CPU usage](#reporting-per-request-cpu-usage)
  * [Reporting system call duration](#reporting-system-call-duration)
  * [Reporting request size](#reporting-request-size)
  * [Reporting a per-request finishing account balance](#reporting-a-per-request-finishing-account-balance)
  * [Reporting process-wide CPU usage](#reporting-process-wide-cpu-usage)
  * [Reporting per-shard memory holdings](#reporting-per-shard-memory-holdings)

<!-- tocstop -->

## Overview

The OpenTelemetry Metrics API supports capturing measurements about

> the execution of a computer program in real time.  The Metrics API is

Reviewer 2: The term "real time" in used through this doc, is there a difference between this and just "time" that is being expressed?

Reviewer 5: Here it looks like it means "concurrent with program execution" and below it looks like it means "clock time". I'd vote to keep these and change the one below.

Author: I've used "run time" and "simultaneously" to express these. PTAL.

designed explicitly for processing raw measurements, generally with
the intent to produce continuous summaries of those measurements, also
in real time.  Hereafter, "the API" refers to the OpenTelemetry
Metrics API.

> The API provides functions for entering raw measurements, through

Author: Yes, in part (😃😃😃). But I've added 
```
The word "enter" as used
here refers to the logical creation of an event through one of the
calling conventions.  This moment of entry (at "run time") defines an
implicit timestamp, which is the wall time an SDK would read from a
clock at that moment.
```

> The API provides functions for entering raw measurements, through

Author: I've replaced "enter" with "capture" everywhere.

> several [calling conventions](TODO: link to user doc) that offer

Reviewer 2: I think it is a great idea to link to the api-metric-user doc here (like the TODO seems to suggest). But, I think we might want to be more explicit about what purpose this doc serves. Based on my reading this is more to provide context and guidance than normative requirements of the API itself, right?

Author: I moved this paragraph to the end of the section to clarify this. Also fixed the TODO.

different levels of performance.  Regardless of calling convention, we
define a _metric event_ as the logical thing that happens when a new
measurement is entered.

Monitoring and alerting systems commonly use the data provided through
metric events, after applying various [aggregations](#aggregations)
and converting into various [exposition formats](#exposition-formats).
However, we find that there are many other uses for metric events,
such as to record aggregated or raw measurements in tracing and
logging systems.  For this reason, [OpenTelemetry requires a
separation of the API from the SDK](library-guidelines.md#requirements),
so that different SDKs can be configured at runtime.

> The word "semantic" or "semantics" as used here refers to _how we give

Reviewer 2: I think this is a great addition and helpful in understanding how to read the rest of the doc. 👍

Author: (I rearranged these paragraphs)

meaning_ to metric events, as they take place under the API.  The term
is used extensively in this document to define and explain these API
functions and how we should interpret them.  As far as possible, the
terminology used here tries to convey the intended semantics, and a
_standard implementation_ will be described below to help us
understand their meaning.  The standard implementation performs
aggregation corresponding to the default interpretation for each kind
of metric event.

### Metric Instruments

A _metric instrument_, of which there are three kinds, is a device for
entering raw measurements into the API.  There are Counter, Measure,
and Observer instruments, each with different semantics and intended
uses, that will be specified here.  All measurements that enter the
API are associated with an instrument, which gives the measurement its
properties.  Instruments are created and defined through calls to a
`Meter` API, which is the user-facing entry point to the SDK.

> Each kinds of metric instrument has its own semantics, briefly

Author: Done.

described as:

- Counter: metric events of this kind _Add_ to a value that is summed over time
- Measure: metric events of this kind _Record_ a value that is aggregated over time

> - Observer: metric events of this kind _Observe_ a coherent set of values at an instant in time.

Author: Done.

An _instrument definition_ describes several properties of the
instrument, including its name and its kind.  The other properties of
a metric instrument are optional, including a description, the unit of
measurement, and several settings that convey additional meaning
(e.g., monotonicity).  An instrument definition is associated with the
events that it produces.

Details about calling conventions for each kind of instrument are
covered in the [user-level API specification](api-metrics-user.md).

### Label sets

_Label_ is the term used to refer to a key-value attribute associated
with a metric event.  Although they are fundamentally similar to [Span
attributes](api-tracing.md#span) in the tracing API, a [label
set](TODO: link to user doc) is given its own type in the Metrics API
(generally: `LabelSet`).  Label sets are a feature of the API meant to

> facilitate re-use, to lower the cost of processing metric events.

Reviewer 2: ```suggestion
facilitate re-use and thereby lowering the cost of processing metric events.
```

Author: Done.

Users are encouraged to re-use label sets whenever possible, as they
may contain a previously encoded representation of the labels.

Users obtain label sets by calling a `Meter` API function.  Each of
the instrument calling conventions detailed in the [user-level API
specification](api-metrics-user.md) accepts a label set.

### Meter Interface

To produce measurements using an instrument, you need an SDK that

> implements the `Meter` API, which consists of a set of instrument

Author: Done.

constructors, functionality related to label sets, and a facility for
entering batches of measurements in a semantically atomic way.

> There is a global `Meter` instance available for use.  Use of the this

Author: I added a description of the Named Meter

> There is a global `Meter` instance available for use.  Use of the this

Author: I wrote
```
There is a global `Meter` instance available for use that facilitates
automatic instrumentation for third-party code.  Use of this instance
allows code to statically initialize its metric instruments, without
explicit dependency injection.  The global `Meter` instance acts as a
no-op implementation until the application explicitly initializes a
global `Meter` by installing an SDK.
```

> instance allows library code that uses it to be automatically enabled

Reviewer 2: This doesn't seem to accurately reflect the operation of the global `Meter` instance. Maybe something like this?

```suggestion
instance allows library code to operate in a no-op fashion until
```

Author: I rewrote this as
```
There is a global `Meter` instance available for use that facilitates
automatic instrumentation for third-party code.  Use of this instance
allows code to statically initialize its metric instruments, without
explicit dependency injection.  The global `Meter` instance acts as a
no-op implementation until the application explicitly initializes a
global `Meter` by installing an SDK.
```

and their data can be examined in different ways.  Using the default
SDK, or an alternate one, we are able to change the interpretation of
metric events at runtime.

Other standard aggregations are available, especially for Measure
instruments, where we are generally interested in a variety of forms
of statistics, such as histogram and quantile summaries.

### Time

Time is a fundamental property of metric events, but not an explicit
one.  Users do not provide explicit timestamps for metric events.
SDKs are discouraged from capturing the current timestamp for each
event (by reading from a clock) unless there is a definite need for
high-precision timestamps.

This non-requirement stems from a common optimization in metrics
reporting, which is to configure metric data collection with a
relatively small period (e.g., 1 second) and use a single timestamp to
describe a batch of exported data, since the loss of precision is
insignificant when aggregating data across minutes or hours of data.

Aggregations are commonly computed over a series of events that fall
into a contiguous region of time, known as the collection interval.

> Since the SDK controls decision to start collection, it is possible to

Author: Done.

form a sufficient basis for expressing nearly all metric data.  But if
the API and SDK are separated, and the SDK can handle any metric event
as it pleases, why not have just one kind of instrument?  How are the
instruments fundamentally different, and why are there only three?

Establishing different kinds of instrument is important because in
most cases it allows the SDK to provide good default functionality,
without requiring alternative behaviors to be configured.  The choice
of instrument determines not only the meaning of the events but also
the name of the function used to report data.  The function
names--`Add()` for Counter instruments, `Record()` for Measure
instruments, and `Observe()` for Observer instruments--help convey the
meaning of these actions.

### Standard implementation

The standard implementation for the three instruments is defined as
follows:

1. Counter.  The `Add()` function accumulates a total for each
distinct label set.  When aggregating over distinct label sets for a
Counter, combine using arithmetic addition and export as a sum.
Depending on the exposition format, sums are exported either as pairs

> of label set and cumulative _difference_ or as pairs of label set and

Author: I've used "delta".

cumulative _total.

2. Measure.  Use the `Record()` function to report events that for
which the SDK will compute summary statistics about the distribution
of values, for each distinct label set.  The summary statistics to use
are determined by the aggregation, but they usually include at least
the sum of values, the count of measurements, and the minimum and
maximum values.  When aggregating distinct Measure events, report
summary statistics of the combined value distribution.  Exposition
formats for summary statistics vary widely, but typically include
pairs of label set and (sum, count, minimum and maximum value).

3. Observer.  Current values are provided by the Observer callback at
the end of each Metric collection period.  When aggregating values

> _for the same label set_, combine using the most-recent value.  When

Author: ```
These values are considered coherent, because measurements from an
Observer instrument in a single collection interval are captured at
the same logical time.  A single callback invocation generates (zero
or more) simultaneous metric events, all sharing an implicit timestamp.
```

aggregating values _for different label sets_, combine the value

> distribution as for Measure instruments.  Export as pairs of label set

Reviewer 5: Here and above it sounds like there are two different kinds of aggregation we have to deal with: (1) over time in the same collection interval and (2) across dimensions. Most of the text above seems to be about (1), but here we're talking about (2) only.

AFAICT the big open question in (2) is how to handle aggregation when one event's label keys are a subset of another's. Without describing aggregators in more detail, I think it's going to be hard to explain the difference between metric instruments here.

I don't mean to suggest moving the spec for aggregators here (and I know it's coming in #347), I just think this might need to be revised after the SDK-side spec is done.

Author: Yes! 

My leading statements in last week's SIG meeting were also on this topic--this is when we get into "Measure-like" and "Counter-like" observers, for example. (Still no CounterMeasure.)

and (sum, count, minimum and maximum value).

We believe that the standard behavior of one of these three
instruments covers nearly all use-cases for users of OpenTelemetry in
terms of the intended semantics.

### Option: Dedicated Measure instrument for timing measurements

> As a language-optional feature, the API may support a dedicated

Reviewer 2: ```suggestion
As a language-optional feature, the API MAY support a dedicated
```

Author: Done.

instrument for reporting timing measurements.  This kind of
instrument, with recommended name `TimingMeasure` (and

> `BoundTimingMeasure`), is semantically equivalent to a Measure

Reviewer 7: I really like this idea. Could we name this `Timer` and `BoundTimer`?

Author: Done.

instrument, and like the Measure instrument supports a `Record()`
function, but the input value to this instrument is in the language's
conventional data type for timing measurements.

For example, in Go the API will accept a `time.Duration`, and in C++

> the API will accept a `std::chrono::duration`.  These advantage of

Reviewer 2: ```suggestion
the API will accept a `std::chrono::duration`.  The advantage of
```

using these instruments is that they use the correct units

> automatically, avoiding the potential for confusion over timing metrics.

Author: I added 
```
Timer instruments MUST capture only the magnitude of the input value
(i.e., an absolute value).  When the user provides a negative value to
the Timer `Record()` function, the captured measurement is the number
stripped of its sign.
```

### Future Work: Option Support

> We are aware of a number of reasons to refine on these types, in order

Reviewer 2: ```suggestion
We are aware of a number of reasons to iterate on these instrument kinds, in order
```

Author: Done.

values is rare in most applications, for example, and it is useful to
offer both a semantic declaration (e.g., "negative values are
meaningless") and a data validation step (e.g., "negative values
should be dropped").
2. Monotonicity support.  When a series of values is known to be
monotonic, it is useful to declare this; this allows us to detect
process resets.

For the most part, these behaviors are not necessary for correctness
within the local process or the SDK, but they are valuable in
down-stream services that use this data.  We look to future work on
this subject.

### Future Work: Configurable Aggregations / View API

The API does not support configurable aggregations, in this
specification.  This is a requirement for OpenTelemetry, but there are
two ways this has been requested.

A _View API_ is defined as an interface to an SDK mechanism that
supports configuring aggregations, including which operator is applied
(sum, p99, last-value, etc.) and which dimensions are used.

1. Should the API user be provided with options to configure specific
views, statically, in the source?
2. Should the View API be a stand-alone facility, able to install

> configurable aggregations, at runtime?

Reviewer 2: +1

> configurable aggregations, at runtime?

Author: 🤷‍♂

Users are encouraged to re-use label sets whenever possible, as they
may contain a previously encoded representation of the labels.

Users obtain label sets by calling a `Meter` API function.  Each of
the instrument calling conventions detailed in the [user-level API
specification](api-metrics-user.md) accepts a label set.

### Meter Interface

To produce measurements using an instrument, you need an SDK that
supports the `Meter` API, which consists of a set of instrument
constructors, functionality related to label sets, and a facility for
reporting batches of measurements in a semantically atomic way.

There is a global `Meter` instance available for use.  Use of the this
instance allows library code that uses it to be automatically enabled
whenever the main application configures an SDK at the global level.

Details about installing an SDK and obtaining a `Meter` are covered in
the [SDK-level API specification](api-metrics-meter.md).

### Aggregations

_Aggregation_ refers to the process of combining a large number of
measurements into exact or estimated statistics about the metric
events that took place during a window of real time, during program

> execution.  Computing aggregations is mainly a subject of the SDK

Reviewer: , with the goal of reducing the amount of data that must be sent to the telemetry collection backend.

specification.

Users do not have a facility in the API to select the aggregation they

> want for particular instruments.  The choice of instrument dictates

Reviewer: Is this just the aspect of the current state of development or the goal? Why wouldn't we want the API to have a way of describing the desired aggregations?

Below we say "Other standard aggregations are available", which feels like they should be part of the API.

semantics and thus gives a default interpretation.  For the standard
implementation:

- Counter instruments use _Sum_ aggregation
- Measure instruments use _MinMaxSumCount_ aggregation
- Observer instruments use _LastValue_ aggregation.

The default Metric SDK specification includes support for configuring
alternative aggregations, so that metric instruments can be repuposed
and their data can be examined in different ways.  Using the default
SDK, or an alternate one, we are able to change the interpretation of
metric events at runtime.

Other standard aggregations are available, especially for Measure
instruments, where we are generally interested in a variety of forms
of statistics, such as histogram and quantile summaries.

### Time

Time is a fundamental property of metric events, but not an explicit
one.  Users do not provide explicit timestamps for metric events.  The

> SDK is welcome to, but not required to capture the current timestamp

Reviewer: "welcome to" seems like too strong. Capturing timestamps is expensive and not necessary for many standard aggregations, which only need timestamps at the end of each aggregation window. I would say that the SDK should only capture timestamps with each measurement when it is necessary for the selected aggregation (are there even real/supported examples when it's necessary?)

Author: I changed it to 

> SDKs are discouraged from capturing the current timestamp for each event (by reading from a clock) unless there is a definite need for high-precision timestamps.

I think one user mentioned wanting to pass metric events through to an existing exporter of some sort that would probably need to know the timestamp, e.g., for a exponential moving average. I suspect it would be better to aggregate over short windows and then compute an EMA, so I think you're right that very few applications call for computing a timestamp per event.

for each event by reading from a clock.

This non-requirement stems from a common optimization in metrics
reporting, which is to configure metric data collection with a
relatively small period (e.g., 1 second) and use a single timestamp to
describe a batch of exported data, since the loss of precision is
insignificant when aggregating data across minutes or hours of data.

Aggregations are commonly computed over a series of events that fall
into a contiguous region of time, known as the collection interval.
Since the SDK controls decision to start collection, it is possible to
collect aggregated metric data while only reading the clock once per
collection interval.

Counter and Measure instruments offer synchronous APIs for entering
measurements.  Metric events from Counter and Measure instruments

> happen when they happen, the moment the SDK receives the function

Reviewer: first "happen" -> "are captured"

requests, broken down by transaction type (e.g., withdrawal, deposit).

Use a Measure instrument to report the current account balance at the
end of each request.  Use a label for the transaction type.

### Reporting process-wide CPU usage

You are interested in reporting the CPU usage of the process as a
whole, which is computed via a (relatively expensive) system call
which returns two values, process-lifetime user and system
cpu-seconds.  It is not necessary to update this measurement
frequently, because it is meant to be used only for accounting
purposes.

A single Observer instrument is recommended for this case, with a
label value to distinguish user from system CPU time.  The Observer
callback will be called once per collection interval, which lowers the
cost of collecting this information.

CPU usage is something that we naturally sum, which raises several
questions.

- Why not use a Counter instrument?  In order to use a Counter instrument, we would need to convert total usage figures into deltas.  Calculating deltas from the previous measurement is easy to do, but Counter instruments are not meant to be used from callbacks.

> - Why not report deltass in the Observer callback?  Observer instruments are meant to be used to observe current values. Nothing prevents reporting deltas with an Observer, but the standard aggregation for Observer instruments is to sum the current value across distinct label sets.  The standard behavior is useful for determining the current rate of CPU usage, but special configuration would be required for an Observer instrument to use Counter aggregation.

Reviewer 8: ```suggestion
- Why not report deltas in the Observer callback?  Observer instruments are meant to be used to observe current values. Nothing prevents reporting deltas with an Observer, but the standard aggregation for Observer instruments is to sum the current value across distinct label sets.  The standard behavior is useful for determining the current rate of CPU usage, but special configuration would be required for an Observer instrument to use Counter aggregation.
```

Author: Fixed.

of the instrumenting library (optionally, the version) when obtaining
a `Meter` implementation, that is meant to be used for identifying
instrumentation produced from that library for such purposes as
disabling instrumentation, configuring aggregation, and applying
sampling policies.  (TODO: refer to the semantic convention on the
reporting library name).

Details about installing an SDK and obtaining a named `Meter` are
covered in the [SDK-level API specification](api-metrics-meter.md).

### Aggregations

_Aggregation_ refers to the process of combining a large number of
measurements into exact or estimated statistics about the metric
events that took place during a window of real time, during program
execution.  Computing aggregations is mainly a subject of the SDK
specification, with the goal of reducing the amount of data that must
be sent to the telemetry collection backend.

Users do not have a facility in the API to select the aggregation they
want for particular instruments.  The choice of instrument dictates
semantics and thus gives a default interpretation.  For the standard
implementation:

- Counter instruments use _Sum_ aggregation

> - Measure instruments use _MinMaxSumCount_ aggregation

Reviewer 4: nit: Not to block this PR, maybe the defaults should be discussed more, I think just SumCount is good enough. Maybe just an issue and associate it with v0.4 milestone?

Author: Issue filed: https://github.com/open-telemetry/opentelemetry-specification/issues/462

Since the SDK controls the decision to start collection, it is possible to
collect aggregated metric data while only reading the clock once per
collection interval.  The default SDK takes this approach.

Counter and Measure instruments offer synchronous APIs for capturing
measurements.  Metric events from Counter and Measure instruments are
captured at the moment they happen, when the SDK receives the
corresponding function call.

The Observer instrument supports an asynchronous API, allowing the SDK
to collect metric data on demand, once per collection interval.  A
single Observer instrument callback can capture multiple metric events
associated with different label sets.  Semantically, by definition,
these observations are captured at a single instant in time, the
instant that they became the current set of last-measured values.

Because metric events are implicitly timestamped, we could refer to a
series of metric events as a _time series_. However, we reserve the
use of this term for the SDK specification, to refer to parts of a
data format that express explicitly timestamped values, in a sequence,
resulting from an aggregation of raw measurements over time.

> ### WithRecommendedKeys declaration on metric instruments

Reviewer 4: I have seconds thoughts on this, maybe file an issue (for v0.4) to discuss if we really need it. Also the argument that we can later add this option can be made here as well :)

Author: #463

A standard feature of metric SDKs is to pre-aggregate metric events
according to a specified set of label keys (i.e., dimensions).  To
perform this task, the SDK must aggregate metric events over the
collection interval: (1) across time, (2) across key dimensions in
_label space_.

When aggregating across spatial dimensions, metric events for
different label sets are combined into an aggregated value for each
distinct "group" of values for the key dimensions.  It means that
measurements are combined for all metric events having the same values
for selected keys, explicitly disregarding any additional labels with
keys not in the set of aggregation keys.  Some exporters are known to
require pre-specifying the label keys used for aggregation (e.g.,

> Prometheus).

Reviewer 4: The Prometheus protocol (exporter) does not require this, only their client requires this for performance reasons to be able to group by a list of label values (see here https://prometheus.io/docs/concepts/data_model). Even OpenMetrics does not require that.

So currently:
* Prometheus - does not require predefined keys
* OpenMetrics - does not require predefined keys
* Statsd/DogStatsd - does not require predefined keys
* Stackdriver - does not require predefined keys (only create MetricDescriptor requires that, but that API is optional)

I am not sure who needs it :( sorry for asking to add this long time ago.

Author: https://github.com/open-telemetry/opentelemetry-specification/issues/463

For example, if `[ak1, ak2]` are the aggregation keys and `[ik1,
ik2]` are the ignored keys, then a metric event having labels
`{ak1=A, ak2=B, ik1=C, ik1=D}` will be combined with a metric
event having labels `{ak1=A, ak2=B, ik1=Y, ik1=Z}` because they
have identical label values for all of the aggregation keys.

> The API provides a `WithKeys` option for the user to declare the

Reviewer: What's the origin of WithKeys function name? Sounds like AggregateBy would be a lot more accurate.

Author: In `api-metrics-user.md` it's defined as "| Recommended label keys | WithRecommendedKeys(list) | Recommended grouping keys for this instrument. |"

(It's called `WithKeys` in the OTel-Go repo, we'll fix that.)

I agree with your opinion, though. I would prefer to make this change (WithRecommendedKeys->AggregateBy) in another PR, not in the 0.3 release.

Author: Let's discuss this name change in v0.4. Resolving this.

recommended aggregation keys when declaring new metric instruments,
intended as the default way to configure an exporter for
pre-aggregation, if it is expected.  Since this is only expected in
some exporters, it is regarded as an option relevant to the exporter,
whether keys configured through `WithRecommendedKeys` are applied for aggregation

> purposes or not.  This allows the user influence the standard

Reviewer 9: ```suggestion
purposes or not.  This allows the user to influence the standard
```

Author: Done.

implementation behavior, especially for exporters that require

> pre-specified aggregation keys.

Reviewer 4: I would say that this is not necessary an exporter property is more like a `View` property :) But it is nit, and as I said after implementing this in Java I think that just removing this it is fine.

Author: https://github.com/open-telemetry/opentelemetry-specification/issues/463

- [Context](context.md) (Span context, Correlation context)
- timestamp (implicit to the SDK)
- instrument definition (name, kind, and semantic options)
- label set (associated key-values)
- value (a signed integer or floating point number)

This format is the result of separating the API from the SDK--a common
representation for metric events, where the only semantic distinction
is the kind of instrument that was specified by the user.

## Three kinds of instrument

Because the API is separated from the SDK, the implementation
ultimately determines how metric events are handled.  Therefore, the
choice of instrument should be guided by semantics and the intended
interpretation.  Here we detail the three instruments and their
individual semantics.

### Counter

Counter instruments are used to capture changes in running sums,
synchronously.  These are commonly used to monitor rates, and they are
sometimes used to capture totals that rise and fall.  An essential
property of Counter instruments is that two events `Add(m)` and
`Add(n)` are semantically equivalent to one event `Add(m+n)`.  This
property means that Counter events can be combined inexpensively, by

> definition.  Counter `Add(0)` events are no-ops, by definition.

Reviewer 4: I would not say this "Counter `Add(0)` events are no-ops, by definition.". If you want to have a Count (number of metric events) aggregation on the CounterInstrument I would expect an `Add(0)` to also be counted.

Author: We'll say that they are "still events". Remove "no-op".

Author: I've written
```
Note that `Add(0)` events are not considered a special case, despite
contributing nothing to a sum.  `Add(0)` events MUST be observed by
the SDK in case non-default aggregations are configured for the
instrument.
```

Observer instruments are used to capture a _current set of values_ at
a point in time.  Observer instruments are asynchronous, with the use
of callbacks allowing the user to capture multiple values per
collection interval.  Observer instruments are not associated with a
Context, by definition.  This means, for example, it is not possible
to associate Observer instrument events with Correlation or Span
context.

Observer instruments capture not only current values, but also
effectively _which label sets are current_ at the moment of
collection.  These instruments can be used to compute probabilities
and ratios, because values are part of a set.

Unlike Counter and Measure instruments, Observer instruments are
synchronized with collection.  There is no aggregation across time for
Observer instruments by definition, only the current set of values is
semantically defined.  Because Observer instruments are activated by
the SDK, they can be effectively disabled at low cost.

These values are considered coherent, because measurements from an
Observer instrument in a single collection interval are captured at
the same logical time.  A single callback invocation generates (zero

> or more) simultaneous metric events, all sharing an implicit timestamp.

Reviewer 4: It is important to mention that a monotonic value is considered to be the sum over the interval [observer_start_time, collection_time] and not [previous_collection_time, current_collection_time]. In other words the value is cumulative  from the beginning and not a delta from the past collection interval. I am proposing this behavior based on the classic example CPU usage (but we can make it the behavior to be delta, but we need to document). Another option is to have an option on the Observer which says cumulative vs delta, but that can be added later, we just need to document the behavior at this moment.

Author: It seems we have 3 kinds of observer, maybe:
- observe a cumulative sum
- observe a delta sum
- observe a measurement value

Author: https://github.com/open-telemetry/opentelemetry-specification/issues/467

We believe that the standard behavior of one of these three
instruments covers nearly all use-cases for users of OpenTelemetry in
terms of the intended semantics.

### Option: Dedicated Measure instrument for timing measurements

As a language-optional feature, the API MAY support a dedicated
instrument for reporting timing measurements.  This kind of
instrument, with recommended name `Timer` (and `BoundTimer`), is
semantically equivalent to a Measure instrument.  Like the Measure
instrument, Timers support a `Record()` function.  The input value to
this instrument is in the language's conventional data type for timing
measurements.

Timer instruments MUST capture only the magnitude of the input value
(i.e., an absolute value).  When the user provides a negative value to
the Timer `Record()` function, the captured measurement is the number
stripped of its sign.

For example, in Go the API will accept a `time.Duration`, and in C++
the API will accept a `std::chrono::duration`.  These instruments
apply the correct units automatically, reducing the potential for

> confusion over timing metric events.

Reviewer 4: Can we please make this `Contrib` for the moment? It is easy to make this LongMeasure and unit `ns` at this point. I am trying to limit the API.

Would be better to have this text somewhere and maybe remove it from here for the moment.

Author: This has a lot of support. We've discussed adding more instrument types, as an alternative. We will pull this from 0.3.

Author: https://github.com/open-telemetry/opentelemetry-specification/issues/464

### Future Work: Option Support

We are aware of a number of reasons to iterate on these
instrumentation kinds, in order to offer:

1. Range restrictions on input data.  Instruments accepting negative values is rare in most applications, for example, and it is useful to offer both a semantic declaration (e.g., "negative values are meaningless") and a data validation step (e.g., "negative values should be dropped").
2. Monotonicity support.  When a series of values is known to be monotonic, it is useful to declare this; this allows us to detect process resets.

> 

Reviewer 4: This is not a correct statement, actually it allows us to calculate correct rates :), the process reset is determine in our data model by having a start and end time associated with every collection interval, and the start time is the one that can determine that - if report cumulative then start_time will be the moment the instrument started to collect, if report deltas resets does not matter :)

Reviewer 4: Some backends like Prometheus used this to determine process restarts that was important to calculate rates (I don't know if restarts were important for other reasons as well, but I know that they need it for rates). Also with OpenMetrics they do support start_time as well to determine restarts.

Author: https://github.com/open-telemetry/opentelemetry-specification/issues/465

For the most part, these behaviors are not necessary for correctness
within the local process or the SDK, but they are valuable in
down-stream services that use this data.  We look to future work on
this subject.

### Future Work: Configurable Aggregations / View API

The API does not support configurable aggregations, in this
specification.  This is a requirement for OpenTelemetry, but there are
two ways this has been requested.

A _View API_ is defined as an interface to an SDK mechanism that
supports configuring aggregations, including which operator is applied
(sum, p99, last-value, etc.) and which dimensions are used.

1. Should the API user be provided with options to configure specific views, statically, in the source?

> 2. Should the View API be a stand-alone facility, able to install configurable aggregations, at runtime?

Reviewer 4: nit: I would probably put this in an issue rather than here.

Author: https://github.com/open-telemetry/opentelemetry-specification/issues/466

Author: I'm placing a link to this in the text.

Since the SDK controls the decision to start collection, it is possible to
collect aggregated metric data while only reading the clock once per
collection interval.  The default SDK takes this approach.

Counter and Measure instruments offer synchronous APIs for capturing
measurements.  Metric events from Counter and Measure instruments are
captured at the moment they happen, when the SDK receives the
corresponding function call.

The Observer instrument supports an asynchronous API, allowing the SDK
to collect metric data on demand, once per collection interval.  A
single Observer instrument callback can capture multiple metric events
associated with different label sets.  Semantically, by definition,
these observations are captured at a single instant in time, the
instant that they became the current set of last-measured values.

Because metric events are implicitly timestamped, we could refer to a
series of metric events as a _time series_. However, we reserve the
use of this term for the SDK specification, to refer to parts of a
data format that express explicitly timestamped values, in a sequence,
resulting from an aggregation of raw measurements over time.

> ### WithKeys declaration on metric instruments

Reviewer: this section is a bit cryptic, could we add an example?

Author: Updated:
```
### WithKeys declaration on metric instruments

A standard feature of metric SDKs is to pre-aggregate metric events
according to a specified set of label keys (i.e., dimensions).  To
perform this task, the SDK must aggregate metric events over the
collection interval: (1) across time, (2) across key dimensions in
_label space_.

When aggregating across spatial dimensions, metric events for
different label sets are combined into an aggregated value for each
distinct "group" of values for the key dimensions.  It means that
measurements are combined for all metric events having the same values
for selected keys, explicitly disregarding any additional labels with
keys not in the set of aggregation keys.  Some exporters are known to
require pre-specifying the label keys used for aggregation (e.g.,
Prometheus).

For example, if `[ak1, ak2]` are the aggregation keys and `[ik1,
ik2]` are the ignored keys, then a metric event having labels
`{ak1=A, ak2=B, ik1=C, ik1=D}` will be combined with a metric
event having labels `{ak1=A, ak2=B, ik1=Y, ik1=Z}` because they
have identical label values for all of the aggregation keys.

The API provides a `WithKeys` option for the user to declare the
recommended aggregation keys when declaring new metric instruments,
intended as the default way to configure an exporter for
pre-aggregation, if it is expected.  Since this is only expected in
some exporters, it is regarded as an option relevant to the exporter,
whether keys configured through `WithKeys` are applied for aggregation
purposes or not.  This allows the user influence the standard
implementation behavior, especially for exporters that require
pre-specified aggregation keys.
```

events produced through an instrument consist of:

- [Context](context.md) (Span context, Correlation context)
- timestamp (implicit to the SDK)
- instrument definition (name, kind, and semantic options)
- label set (associated key-values)
- value (a signed integer or floating point number)

This format is the result of separating the API from the SDK--a common
representation for metric events, where the only semantic distinction
is the kind of instrument that was specified by the user.

## Three kinds of instrument

Because the API is separated from the SDK, the implementation
ultimately determines how metric events are handled.  Therefore, the
choice of instrument should be guided by semantics and the intended
interpretation.  Here we detail the three instruments and their
individual semantics.

### Counter

Counter instruments are used to capture changes in running sums,
synchronously.  These are commonly used to monitor rates, and they are
sometimes used to capture totals that rise and fall.  An essential
property of Counter instruments is that two `Add(1)` events are
semantically equivalent to one `Add(2)` event--`Add(m)` and `Add(n)`

> is equivalent to `Add(m+n)`.  This property means that Counter events

Reviewer: Can simplify:

An essential property of Counter instruments is that two  events `Add(m)` and `Add(n)` are semantically equivalent to one event `Add(m+n)`.

Author: Done.

can be combined inexpensively, by definition.

Labels associated with Counter instrument events can be used to
compute rates and totals from the instrument, over selected

> dimensions.  Counter `Add(0)` events are no-ops, by definition.

Reviewer: > Counter `Add(0)` events are no-ops, by definition.

this belongs to the end of previous paragraph.

Author: Done.

### Measure

Semantically, metric events from Measure instruments are independent,
meaning they cannot be combined naturally, as with Counters.  Measure
instruments are used to capture many kinds of information,
synchronously, and are recommended for all cases that reflect an event
in the application where the additive property of Counter instruments
does not apply.

Labels associated with Measure instrument events can be used to
compute information about the distribution of values from the
instrument, over selected dimensions.  When aggregating Measure
events, the output statistics are expected to reflect the combined
data set.

> 

Reviewer: I would add:

It is worth noting that Counter instruments are a special case of Measure instrument with Sum-style aggregation. However, considering that Counters are used extensively in the existing instrumentations, is it extracted to a top-level instrument kind for user convenience.

Author: I don't think about it this way. I'd say that Counter and Measure are equally special cases of Synchronous Instrument, where Counter uses Counter-style aggregation (i.e., Sum) and Measure uses Measure-style aggregation (i.e., MinMaxSumCount). The difference is you said "instrument with Sum-style aggregation" and I'm saying "instrument with the intention of Sum-style aggregation". Anyway I added:

```
Counter instruments can be seen as special cases of Measure
instruments with the additive property described above and a
more-specific verb to improve readability (i.e., "Add" instead of
"Record").  Counter instruments are special cases of Measure
instruments in that they only preserve a Sum, by default, and no other
summary statistics.
```

the API and SDK are separated, and the SDK can handle any metric event
as it pleases, why not have just one kind of instrument?  How are the
instruments fundamentally different, and why are there only three?

Establishing different kinds of instrument is important because in
most cases it allows the SDK to provide good default functionality,
without requiring alternative behaviors to be configured.  The choice
of instrument determines not only the meaning of the events but also
the name of the function used to report data.  The function
names--`Add()` for Counter instruments, `Record()` for Measure
instruments, and `Observe()` for Observer instruments--help convey the
meaning of these actions.

### Standard implementation

The standard implementation for the three instruments is defined as
follows:

1. Counter.  The `Add()` function accumulates a total for each
distinct label set.  When aggregating over distinct label sets for a
Counter, combine using arithmetic addition and export as a sum.
Depending on the exposition format, sums are exported either as pairs
of label set and cumulative _delta_ or as pairs of label set and
cumulative _total_.

> 2. Measure.  Use the `Record()` function to report events that for

Reviewer: ```suggestion
2. Measure.  Use the `Record()` function to report events for
```

Author: Done.

processed by the instrument.  Counters are useful when only the sum is
interesting.  Measures are useful when the sum and any other kind of
summary information about the individual values are of interest.

If only the sum is of interest, use a Counter instrument.

If you are interested in any other kind of summary value or statistic,
such as mean, median and other quantiles, or minimum and maximum
value, use a Measure instrument.  Measure instruments are used to
report any kind of measurement that is not typically expressed as a
rate or as a total sum.

### Observer instruments

Observer instruments are recommended for reporting measurements about
the state of the program periodically.  These expose current
information about the program itself, not related to individual events
taking place in the program.  Observer instruments are reported
outside of a context, thus do not have an effective span context or
correlation context.

Observer instruments are meant to be used when measured values report
on the current state of the program, as opposed to an event or a
change of state in the program.

## Examples

### Reporting bytes read and written

> You wish to monitor the number of bytes read and written from a

Reviewer: ```suggestion
You wish to monitor the total number of bytes read and written from a
```

Consider splitting this example into "total bytes" and "bytes per message". Even though they both use the exact same values to record events, the former uses Counter, but the latter uses Measure.

Overall, this is actually a confusing example because one could argue that the instrumentation should always use Measure in this case, as the more agnostic instrument, and the operator can choose whether they want to report totals or a per-message distribution (or both).

Author: I've written:
```
### Reporting total bytes read

You wish to monitor the total number of bytes read from a messaging
server that supports several protocols.  The number of bytes read
should be labeled with the protocol name and aggregated in the
process.

This is a typical application for the Counter instrument.  Use one
Counter for capturing the number bytes read.  When handling a request,
compute a LabelSet containing the name of the protocol and potentially
other useful labels, then call `Add()` with the same label set and the
number of bytes read.

To lower the cost of this reporting, you can `Bind()` the
instrument with each of the supported protocols ahead of time and
avoid computing the label set for each request.

### Reporting total bytes read and bytes per request

You wish to monitor the total number of bytes read as well as the
number of bytes read per request, to have observability into total
traffic as well as typical request size.  As with the example above,
these metric events should be labeled with a protocol name.

This is a typical application for the Measure instrument.  Use one
Measure for capturing the number of bytes per request.  A sum
aggregation applied to this data yields the total bytes read; other
aggregations allow you to export the minimum and maximum number of
bytes read, as well as the average value, and quantile estimates.

In this case, the guidance is to create a single instrument.  Do not
create a Counter instrument to export a sum when you want to export
other summary statistics using a Measure instrument.
```

messaging server that supports several protocols.  The number of bytes
read and written should be labeled with the protocol name and
aggregated in the process.

This is a typical application for the Counter instrument.  Use one
Counter for bytes read and one Counter for bytes written.  When
handling a request, compute a LabelSet containing the name of the
protocol and potentially other useful labels, then call `Add()` twice
with the same label set and the number of bytes read and written.

To lower the cost of this reporting, you can `Bind()` the
instrument with each of the supported protocols ahead of time and
avoid computing the label set for each request.

### Reporting per-request CPU usage

> Suppose you have a way to measure the CPU usage of processing an

Reviewer: Don't  like the wording of this example. You're talking about wall clock time of the request (ie latency), not the CPU usage (which is quite difficult to measure accurately _per request_).

Author: I removed the example.

callbacks.
- Why not report deltass in the Observer callback?  Observer
instruments are meant to be used to observe current values. Nothing
prevents reporting deltas with an Observer, but the standard
aggregation for Observer instruments is to sum the current value
across distinct label sets.  The standard behavior is useful for
determining the current rate of CPU usage, but special configuration
would be required for an Observer instrument to use Counter
aggregation.

### Reporting per-shard memory holdings

Suppose you have a widely-used library that acts as a client to a
sharded service.  For each shard it maintains some client-side state,
holding a variable amount of memory per shard.

Observe the current allocation per shard using an Observer instrument
with a shard label.  These can be aggregated across hosts to compute
cluster-wide memory holdings by shard, for example, using the standard
aggregation for Observers, which sums the current value across
distinct label sets.

### Reporting number of active requests

Suppose your server maintains the count of active requests, which
rises and falls as new requests begin and end processing.

> Observer the number of active requests periodically with an Observer

Reviewer 3: ```suggestion
Observe the number of active requests periodically with an Observer
```

Author: Fixed.

> 		# Metrics API

Reviewer 9: ```suggestion
# Metrics API
```

callbacks allowing the user to enter multiple values per collection
interval.  

Observer instruments enter not only current values, but also
effectively _which label sets are current_ at the moment of
collection.  These instruments can be used to compute probabilities
and ratios, because values are part of a set.

Unlike Counter and Measure instruments, Observer instruments are
synchronized with collection.  There is no aggregation across time for
Observer instruments by definition, only the current set of values is
semantically defined.

These values are considered coherent, because measurements from an
Observer instrument in a single collection interval are considered
simultaneous.  The set of measurements entered through one callback
invocation implicitly share a timestamp.

#### Range options

Observer instruments are Absolute by default, meaning `Observe()`
logically accepts only non-negative values.  Absolute observers are
commonly used for measuring things like cpu-load or queue sizes.

Unrestricted Observer instruments can be used for quantities that can
be negative, such as account balance or degrees Celsius.

> Observer instruments can be declared as _Monotonic_, a special kind of

Reviewer 3: Below it says that it could also apply for Measure instruments.

Author: This was a stale revision, sorry I added "WIP" too late. It's removed in the current draft.

We recognize that the standard behavior of the three instruments does
not cover all use-cases perfectly.  There is a natural tension between
offering dedicated metric instruments for every distinct metric
application and combining use-cases, generalizing semantics to reduce
the API surface area.  We could have define more than or fewer than
three kinds of instrument; we have three because these seem like
enough.  Where uncommon use-cases call for a non-standard
implementation configuration (e.g., a Measure instrument configured
with last-value aggregation), we accept that users will be required to
provide additional input on how to view certain metric data.

### Optional semantic restrictions

The instruments support optional declarations that indicate
restrictions on the valid range of inputs.  There are two options, one
to indicate whether the value is signed or not, the other to indicate
monotonicity.  These options are meant to be used as a signal to the
observability system, since they impact the way these data are exposed
to users.

In both cases, the optional restriction does not change the semantics
of the instrument.  The options are independent, both can be

> meaningfully set on any instrument kind.

Reviewer 9: Monotonic does not make sense on Counters.
Line 309 says it only applies to Observers, line 427 says it only applies to Measures and Observers.

> meaningfully set on any instrument kind.

Reviewer 9: Monotonic does not make sense on Counters.
Line 309 says it only applies to Observers, line 427 says it only applies to Measures and Observers.

Author: Option support has been removed, per today's SIG.

signed inputs.

When an instrument is absolute (i.e., accepts non-negative updates),
we know that the sum can be used to express a rate automatically.
This is true for all kinds of instrument.

When exporting measure values as a histogram, for example, knowing the
instrument is absolute facilitates the use of logarithmic buckets
(which are difficult to use when the input range spans zero).

Absolute behavior is the default for all instrument kinds. The
Non-Absolute option is supported for all instrument kinds.

Because this is a simple property for the SDK to test, the
specification recommends that SDKs SHOULD reject metric events for
absolute instruments when negative values are used, and instead issue
a warning to the user.

#### Monotonic vs. Non-Monotonic

Monotonic refers to whether an instrument only accepts values that are
greater than or equal to the previously recorded value.  Non-monotonic
instruments are those which accept any change in the value, positive
or negative.

Absolute-valued counters are naturally monotonic, so that Absolute and
Monotonic have the same interpretation for Counter instruments.

> Measure and Observer instruments may be declared as monotonic, however

Reviewer 3: Above https://github.com/open-telemetry/opentelemetry-specification/pull/430/files#r375837745 it  says it applies only for Observer instruments.
